lib/luaxtable.c

```diff
@@ -32,48 +14,169 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret, int fallback)
 {
-	return true;
+	int ret = -1;
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		goto err;
+	}
+	if (nret == 1) {
+		ret = luaL_optinteger(L, -1, 0);
+		ret = ret < 0 ? fallback : ret;
+	}
+err:
+	return ret;
```

Comment:
here we are not returning fallback in case of error.. is it expected?

Reply:
I should probable initialise `ret = fallback` instead of `ret = -1`, It should handle all the edge cases.

Reply:
in that case, I think you don't need `ret` you can return fallback directly on `err: `, it might be clearer..

Reply:
Ok

Reply:
Ah but, I will have to add another return statement above `err:`

Reply:
So it would be : 
```c
if (nret == 1) {
		ret = luaL_optinteger(L, -1, 0);
		ret = ret < 0 ? fallback : ret;
	}
	return ret;
err:
	return fallback;
```